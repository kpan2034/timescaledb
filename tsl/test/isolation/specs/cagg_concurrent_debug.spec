# This file and its contents are licensed under the Timescale License.
# Please see the included NOTICE for copyright information and
# LICENSE-TIMESCALE for a copy of the license.

#
# Setup prior to every permutation.
#
# We define a function 'cagg_bucket_count' to get the number of
# buckets in a continuous aggregate.  We use it to verify that there
# aren't any duplicate buckets/rows inserted into the materialization
# hypertable after concurrent refreshes. Duplicate buckets are
# possible since there is no unique constraint on the GROUP BY keys in
# the materialized hypertable.
#
setup
{
-- create a base table for sensor data
CREATE TABLE IF NOT EXISTS sensor_data (
    time TIMESTAMPTZ NOT NULL,
    sensor_id INTEGER NOT NULL,
    temperature DOUBLE PRECISION,
    humidity DOUBLE PRECISION,
    pressure DOUBLE PRECISION
);

SELECT create_hypertable('sensor_data', 'time',
    chunk_time_interval => INTERVAL '1 day'
);

INSERT INTO sensor_data
SELECT
    timestamp '2025-01-01' + (i * INTERVAL '5 minutes') AS time,
    (i % 5) + 1 AS sensor_id,
    15 + 15 * random() AS temperature,
    30 + 60 * random() AS humidity,
    980 + 40 * random() AS pressure
FROM generate_series(0, 8640) AS i;

CREATE MATERIALIZED VIEW sensor_hourly_avg
WITH (timescaledb.continuous) AS
SELECT
    time_bucket('1 hour', time) AS bucket,
    sensor_id,
    AVG(temperature) AS avg_temperature,
    AVG(humidity) AS avg_humidity,
    AVG(pressure) AS avg_pressure,
    COUNT(*) AS reading_count
FROM sensor_data
GROUP BY bucket, sensor_id
WITH NO DATA;

    CREATE OR REPLACE FUNCTION lock_cagg(cagg name) RETURNS void AS $$
    DECLARE
      mattable text;
    BEGIN
      SELECT format('%I.%I', user_view_schema, user_view_name)
      FROM _timescaledb_catalog.continuous_agg
      WHERE user_view_name = cagg
      INTO mattable;
      EXECUTE format('LOCK table %s IN ROW EXCLUSIVE MODE', mattable);
    END
    $$ LANGUAGE plpgsql;
}

setup
{
CALL refresh_continuous_aggregate('sensor_hourly_avg', NULL, NULL);
}

setup
{
BEGIN;
INSERT INTO sensor_data (time, sensor_id, temperature, humidity,pressure ) VALUES (now()-'1 day'::interval, 2, 20,30,1000);
COMMIT;
BEGIN;
INSERT INTO sensor_data (time, sensor_id, temperature, humidity,pressure ) VALUES (now()-'1 month'::interval, 2, 20,30,1000);
COMMIT;
}

teardown {
    DROP TABLE sensor_data CASCADE;
}

session "R1"
setup
{
    SET SESSION client_min_messages = 'DEBUG1';
}
step "R1_refresh"
{
    CALL refresh_continuous_aggregate('sensor_hourly_avg', '2 months'::interval, '1 week'::interval);
}

session "R2"
setup
{
    SET SESSION client_min_messages = 'DEBUG1';
}
step "R2_refresh"
{

    CALL refresh_continuous_aggregate('sensor_hourly_avg', '1 week'::interval, '1 hour'::interval);
}

session "L"
step debug_waitpoint_enable
{
    SELECT debug_waitpoint_enable('invalidation_process_hypertable_log');
}

step debug_waitpoint_release
{
    SELECT debug_waitpoint_release('invalidation_process_hypertable_log');
}

session "S"
step count_bucket
{
    SELECT count(bucket) FROM sensor_hourly_avg;
}

permutation "debug_waitpoint_enable" "R1_refresh" "R2_refresh" "debug_waitpoint_release" "count_bucket" "R2_refresh" "count_bucket"
