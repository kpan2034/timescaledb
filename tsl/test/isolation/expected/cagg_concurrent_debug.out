Parsed test spec with 4 sessions

starting permutation: debug_waitpoint_enable R1_refresh R2_refresh debug_waitpoint_release count_bucket R2_refresh count_bucket
step debug_waitpoint_enable: 
    SELECT debug_waitpoint_enable('invalidation_process_hypertable_log');

debug_waitpoint_enable
----------------------
                      
(1 row)

R1: LOG:  statement: 
    CALL refresh_continuous_aggregate('sensor_hourly_avg', '2 months'::interval, '1 week'::interval);

step R1_refresh: 
    CALL refresh_continuous_aggregate('sensor_hourly_avg', '2 months'::interval, '1 week'::interval);
 <waiting ...>
R2: LOG:  statement: 

    CALL refresh_continuous_aggregate('sensor_hourly_avg', '1 week'::interval, '1 hour'::interval);

step R2_refresh: 

    CALL refresh_continuous_aggregate('sensor_hourly_avg', '1 week'::interval, '1 hour'::interval);
 <waiting ...>
step debug_waitpoint_release: 
    SELECT debug_waitpoint_release('invalidation_process_hypertable_log');

debug_waitpoint_release
-----------------------
                       
(1 row)

R1: DEBUG:  continuous aggregate refresh (individual invalidation) on "sensor_hourly_avg" in window [ Sat May 10 16:00:00 2025 PDT, Thu Jul 03 15:00:00 2025 PDT ]
R1: LOG:  deleted 0 row(s) from materialization table "_timescaledb_internal._materialized_hypertable_X"
R1: DEBUG:  building index "_hyper_X_X_chunk__materialized_hypertable_X_bucket_idx" on table "_hyper_X_X_chunk" serially
R1: DEBUG:  index "_hyper_X_X_chunk__materialized_hypertable_X_bucket_idx" can safely use deduplication
R1: DEBUG:  building index "_hyper_X_X_chunk__materialized_hypertable_X_sensor_id_bucket_i" on table "_hyper_X_X_chunk" serially
R1: DEBUG:  index "_hyper_X_X_chunk__materialized_hypertable_X_sensor_id_bucket_i" can safely use deduplication
R1: LOG:  inserted 1 row(s) into materialization table "_timescaledb_internal._materialized_hypertable_X"
step R1_refresh: <... completed>
R2: DEBUG:  cannot lock "sensor_hourly_avg" materialization invalidation log [ -infinity, -infinity ], skipping
R2: DEBUG:  cannot lock "sensor_hourly_avg" materialization invalidation log [ Fri Jan 31 01:00:00 2025 PST, infinity ], skipping
R2: NOTICE:  continuous aggregate "sensor_hourly_avg" is already up-to-date
step R2_refresh: <... completed>
step count_bucket: 
    SELECT count(bucket) FROM sensor_hourly_avg;

count
-----
 3602
(1 row)

R2: LOG:  statement: 

    CALL refresh_continuous_aggregate('sensor_hourly_avg', '1 week'::interval, '1 hour'::interval);

R2: DEBUG:  hypertable 1 existing watermark >= new invalidation threshold 1752181200000000 1752181200000000
R2: DEBUG:  continuous aggregate refresh (individual invalidation) on "sensor_hourly_avg" in window [ Thu Jul 03 16:00:00 2025 PDT, Thu Jul 10 14:00:00 2025 PDT ]
R2: LOG:  deleted 0 row(s) from materialization table "_timescaledb_internal._materialized_hypertable_X"
R2: DEBUG:  building index "_hyper_X_X_chunk__materialized_hypertable_X_bucket_idx" on table "_hyper_X_X_chunk" serially
R2: DEBUG:  index "_hyper_X_X_chunk__materialized_hypertable_X_bucket_idx" can safely use deduplication
R2: DEBUG:  building index "_hyper_X_X_chunk__materialized_hypertable_X_sensor_id_bucket_i" on table "_hyper_X_X_chunk" serially
R2: DEBUG:  index "_hyper_X_X_chunk__materialized_hypertable_X_sensor_id_bucket_i" can safely use deduplication
R2: LOG:  inserted 1 row(s) into materialization table "_timescaledb_internal._materialized_hypertable_X"
step R2_refresh: 

    CALL refresh_continuous_aggregate('sensor_hourly_avg', '1 week'::interval, '1 hour'::interval);

step count_bucket: 
    SELECT count(bucket) FROM sensor_hourly_avg;

count
-----
 3603
(1 row)

